portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java

```diff
@@ -326,6 +336,62 @@ public boolean shouldDropIndex(Collection<String> columnNames) {
 
 	}
 
+	protected static <T, E1 extends Exception, E2 extends Exception> void
+			concurrentUpgrade(
+				UnsafeSupplier<T, E1> unsafeSupplier,
+				UnsafeConsumer<T, E2> unsafeConsumer)
+		throws E1, E2 {
+
+		Objects.requireNonNull(unsafeSupplier);
+		Objects.requireNonNull(unsafeConsumer);
+
+		ExecutorService executorService = Executors.newWorkStealingPool();
+
+		ThrowableCollector throwableCollector = new ThrowableCollector();
+
+		List<Future<Void>> futures = new ArrayList<>();
+
+		try {
+			long companyId = CompanyThreadLocal.getCompanyId();
+
+			T next = null;
+
+			while ((next = unsafeSupplier.get()) != null) {
+				T current = next;
+
+				Future<Void> future = executorService.submit(
+					() -> {
+						try (SafeCloseable safeCloseable =
+								CompanyThreadLocal.lock(companyId)) {
+
```

Comment:
I guess you lock the ThreadLocal here because you are not using BaseUpgradeCallable in https://github.com/achaparro/liferay-portal/pull/981/files#diff-2ef28567fdf5e3d5d374380d3bbd91269f5d0542e13bfc97d705185c051e133eR74 so you need it. 

It would be better to have one single point where we lock the CompanyThreadLocal, we can:
- Force everyone to use BaseUpgradeCallable
or
- Remove the use of BaseUpgradeCallable in all cases and put the lock in the place you suggested

Reply:
I agree, I think here is better since it doesn't require extending anything.

Reply:
But it requires us to create objects like journalArticleInfo which can be more confusing than the callable. Maybe we can generate an array of objects or something like that so we don't have to define those classes. Although it is less verbose and it will require casting the values.

Reply:
It doesn't require it, see 5b79702 and 7c77b73.